nx/lib/nx/shape.ex

```diff
@@ -556,6 +556,78 @@ defmodule Nx.Shape do
     end
   end
 
+  @doc """
+  Flattens the given axes of the given input shape into a
+  single axis.
+
+  ## Examples
+
+      iex> Nx.Shape.flatten({1, 2, 3}, [nil, nil, nil], nil)
+      {{6}, [nil]}
+
+      iex> Nx.Shape.flatten({1, 2, 3}, [:batch, nil, nil], [1, 2])
+      {{1, 6}, [:batch, nil]}
+
+      iex> Nx.Shape.flatten({1, 2, 3}, [nil, nil, nil], [])
+      {{1, 2, 3}, [nil, nil, nil]}
+
+  ### Error cases
+
+      iex> Nx.Shape.flatten({1, 2, 3}, [:batch, nil, nil], [0, 2])
+      ** (ArgumentError) flatten axes must be consecutive
+  """
+  def flatten(shape, names, axes)
+
+  def flatten({}, _, _), do: {{}, []}
```

Comment:
Shouldn't we raise if you pass axes and the shape does not have said axes? So I think we still need to call `normalize_axes` on this? Perhaps:

```suggestion
  def flatten({}, [], []), do: {{}, []}
```

But perhaps not even worth having this branch because it is not common to flatten scalars?